### PR TITLE
[MB-1996] Add placeholder values and comments to envrc.local.template

### PR DESCRIPTION
**Pull request by Ronolibert**
## Description

Define out what is the bare minimum to get the app to work given the required environment variables

## Reviewer Notes
- Is this enough? Or should the template values be actual valid values
- For `LOGIN_GOV_SECRET_KEY`, `MOVE_MIL_DOD_TLS_CERT` and `MOVE_MIL_DOD_TLS_KEY` I'm not sure if we want real values because the pre-commit is stopping me from committing actual secret keys
- If we want real values, I'm not sure how to values for the login_gov_secret_key, and the certs. I just know what error messages the system gave when they were missing and I put those as the comments

## Setup
- Re-clone mymove project
- Create a `.envrc.local` file
- Copy paste the template from this PR
- Fill in the env variables given the comments
- Run `make server_run`
## Code Review Verification Steps
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References
* [Jira story](https://dp3.atlassian.net/browse/MB-1996) for this change

### CODE REVIEW

File: .envrc.local.template [hunk 1]
```diff
@@ -1,12 +1,24 @@
...
-export EIA_KEY=replace
-export CLOUD_FRONT_PRIVATE_KEY=replace
-export CLOUD_FRONT_KEY_ID=replace
\ No newline at end of file
+# any string length 32
+export DPS_AUTH_COOKIE_SECRET_KEY=thisisakeywithalengthof32charsss
+# any string length 64
+export CSRF_AUTH_KEY=thisisakeywithalengthof64charsssthisisakeywithalengthof64charsss
+# any string of any length
+export HERE_MAPS_APP_ID=placeholder
```
Comment (chrisgilmerproj) [inline .envrc.local.template:12]: A good idea would be to include the link for HERE maps in the comment: https://www.here.com/

File: .envrc.local.template [hunk 2]
```diff
@@ -1,12 +1,24 @@
...
+# any string of any length
+export HERE_MAPS_APP_CODE=placeholder
+# any string of any length
+export GEX_BASIC_AUTH_PASSWORD=placeholder
+# any string of any length
+export DPS_AUTH_SECRET_KEY=placeholder
+# any string of any length
+export EIA_KEY=placeholder
```
Comment (chrisgilmerproj) [inline .envrc.local.template:20]: I believe they have to get an account at https://www.eia.gov/ for the API.

File: .envrc.local.template [hunk 3]
```diff
@@ -1,12 +1,24 @@
...
+# any string of any length
+export CLOUD_FRONT_PRIVATE_KEY=placeholder
+# any string of any length
+export CLOUD_FRONT_KEY_ID=placeholder
```
Comment (chrisgilmerproj) [inline .envrc.local.template:24]: Turns out they shouldn't need CloudFront to run the app. Can you modify the `.envrc` to remove the `require` for these two items? In fact, in `.envrc` where it discusses `To use cdn` you should just add a comment that these two things will be required. https://github.com/transcom/mymove/blob/master/.envrc#L186

Then you can remove these from this file.

File: .envrc.local.template [hunk 4]
```diff
@@ -1,12 +1,24 @@
...
+# any string length 64
+export CSRF_AUTH_KEY=thisisakeywithalengthof64charsssthisisakeywithalengthof64charsss
+# any string of any length
+export HERE_MAPS_APP_ID=placeholder
+# any string of any length
+export HERE_MAPS_APP_CODE=placeholder
+# any string of any length
+export GEX_BASIC_AUTH_PASSWORD=placeholder
+# any string of any length
+export DPS_AUTH_SECRET_KEY=placeholder
```
Comment (chrisgilmerproj) [inline .envrc.local.template:18]: Can you move this DPS variable up with the other DPS variable? Just keep them together :)

File: .envrc.local.template [hunk 5]
```diff
@@ -1,12 +1,24 @@
+# Key must be PEM encoded PKCS1 https://tools.ietf.org/html/rfc7468#section-2
 export LOGIN_GOV_SECRET_KEY=replace
```
Comment (chrisgilmerproj) [inline .envrc.local.template:2]: Mention that this is connected to https://login.gov

File: .envrc.local.template [hunk 6]
```diff
@@ -1,12 +1,24 @@
...
+# certificate PEM block https://tools.ietf.org/html/rfc7468#section-2
 export MOVE_MIL_DOD_TLS_CERT=replace
+# certificate PEM block https://tools.ietf.org/html/rfc7468#section-2
```
Comment (chrisgilmerproj) [inline .envrc.local.template:5]: If we don't provide keys here then let's give them a command line tool to create them. Probably some `ssh` command.

Comment (Ronolibert) [reply .envrc.local.template:5]: Do you have any resources on how to create these values? The link to those docs are super academic for me and hard to parse

Comment (Ronolibert) [reply .envrc.local.template:5]: I can make a separate card for the cmd line tool if that's alright

Comment (chrisgilmerproj) [reply .envrc.local.template:5]: I think in this case you can follow this blog post here: https://medium.com/@hugomejia74/how-to-create-your-own-ssl-certificate-authority-for-local-https-development-3b97573c7bb5. You don't have to write a cmd line tool, they already exist (openssl, ssh). 

It's important to know what these values are for. The DoD created for us a TLS certificate and key to use for our application when it hosts `*.move.mil`, which is any of our sub-domains. The certificate they gave us is the CA (Certificate Authority) and we use the `MOVE_MIL_DOD_CA_CERT` as the env var for it. In our case the public side of that is stored in `config/tls/dod-sw-ca-54.pem`. So whatever values we use for `MOVE_MIL_DOD_TLS_CERT` and `MOVE_MIL_DOD_TLS_KEY` need to be derived from the `MOVE_MIL_DOD_CA_KEY` which we don't have (only the DoD has it).

We actually have a parallel set of certs already that do this. Maybe you can use them! The files are:

- config/tls/devlocal-ca.pem (try this for MOVE_MIL_DOD_CA_CERT)
- config/tls/devlocal-https.key (try this for MOVE_MIL_DOD_TLS_KEY)
- config/tls/devlocal-https.pem (try this for MOVE_MIL_DOD_TLS_CERT)

like this:

```
MOVE_MIL_DOD_CA_CERT=$(cat "${MYMOVE_DIR}"/config/tls/devlocal-ca.pem)
MOVE_MIL_DOD_TLS_KEY=$(cat "${MYMOVE_DIR}"/config/tls/devlocal-https.key)
MOVE_MIL_DOD_TLS_CERT=$(cat "${MYMOVE_DIR}"/config/tls/devlocal-https.cer)
```

The files `devlocal-https.{key,pem}` were created with the tool at `scripts/generate-devlocal-cert` (this is also how we made files like `config/tls/devlocal-mtls.{key,pem}`. 

So if you can't re-use the files I put above then the steps are:

1. Create a new CA cert
2. Use `scripts/generate-devlocal-cert` and modify it to derive a new key and pem from the new CA cert
3. Copy all the values into the `.envrc.local.template`
4. Delete the files you created earlier

**Comment by Ronolibert** [comment]
@chrisgilmerproj I improved the comments with those links and I was also able to use the local certs that the script generates and it worked fine! Let me know if the changes are fine

File: .envrc [hunk 7]
```diff
@@ -255,10 +255,6 @@ export SERVE_API_GHC=true
 export SERVE_API_PRIME=true
 export SERVE_API_SUPPORT=true
 
-# CDN
-require CLOUD_FRONT_PRIVATE_KEY "See 'chamber read app-devlocal cloud_front_private_key'"
-require CLOUD_FRONT_KEY_ID "See 'chamber read app-devlocal cloud_front_key_id'"
-
```
Comment (chrisgilmerproj) [inline .envrc:261]: You'll want to add a comment referring to these up near the comment about `cdn`. Just so folks know they'll need to set these if they set `STORAGE_BACKEND=cdn`.

File: .envrc.local.template (hunk 6 above)
Comment (chrisgilmerproj) [reply .envrc.local.template:5]: So you can't use the `devlocal-mtls` in this example because its very likely that will cause an issue with anyone using this templated file. I suggested above the steps to make new temporary files to use, copy the values into this file, and then delete them. What you copied in here was just my example to show you how it would work.

File: .envrc (hunk 7 above)
Comment (Ronolibert) [reply .envrc:261]: Good catch!

File: .envrc.local.template (hunk 6 above)
Comment (Ronolibert) [reply .envrc.local.template:5]: Ah I see, I thought this would be fine since we also use those example files for postman. Will do the other approach

File: .envrc (hunk 7 above)
Comment (chrisgilmerproj) [reply .envrc:261]: Love your fix! Thanks.

File: .envrc.local.template (hunk 6 above)
Comment (chrisgilmerproj) [reply .envrc.local.template:5]: I think it might be helpful to know that we can't add the devlocal-ca certs twice, which is what would happen in development with this approach. Because we load the DoD CA first and then we load the Devlocal CA. So what you unintentionally did is add the Devlocal CA and then the Devlocal CA, which can cause issues.

**Comment by chrisgilmerproj** [comment]
@Ronolibert - here's a script I pulled together to make the devlocal ca stuff easier. It's a bit of a hack but it appears to work. It's the values of the output files that you'll end up using in the `.envrc.local.template`.

```sh
#! /usr/bin/env bash

CA_KEY=temp-ca.key
CA_PEM=temp-ca.pem
CA_SRL=temp-ca.srl

# Generate the CA private key
openssl genrsa -out "${CA_KEY}" 2048

# Generate the CA root certificate
# Default subject fields
C="US"
ST="DC"
L="Washington"
CN="localhost"
openssl req -new -key "${CA_KEY}" -x509 -days 3652 -out "${CA_PEM}" -subj "/C=$C/ST=$ST/L=$L/O=$O/OU=$OU/CN=$CN"

# Generate devlocal cert
DEVLOCAL_CER=temp-devlocal.cer
DEVLOCAL_KEY=temp-devlocal.key
DEVLOCAL_CSR=temp-devlocal.csr

openssl req -nodes -new -keyout "${DEVLOCAL_KEY}" -out "${DEVLOCAL_CSR}" -subj "/C=$C/ST=$ST/L=$L/O=$O/OU=$OU/CN=$CN"
openssl x509 -req -in "${DEVLOCAL_CSR}" -CA "${CA_PEM}" -CAkey "${CA_KEY}" -CAcreateserial -out "${DEVLOCAL_CER}" -days 3652 -sha256
echo -n "SHA256 digest: "
openssl x509 -outform der -in "${DEVLOCAL_CER}" | openssl dgst -sha256

# Cleanup
rm -f "${DEVLOCAL_CSR}"
rm -f "${CA_SRL}"
```
